providers/adfs.go

```diff
@@ -0,0 +1,81 @@
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"net/url"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+)
+
+// ADFSProvider represents an ADFS based Identity Provider
+type ADFSProvider struct {
+	*OIDCProvider
+}
+
+var _ Provider = (*OIDCProvider)(nil)
+
+const (
+	adfsProviderName = "Adfs"
+	adfsDefaultScope = "openid email profile"
+)
+
+// NewADFSProvider initiates a new AdfsProvider
+func NewADFSProvider(p *OIDCProvider) *ADFSProvider {
+
+	defaultScope := adfsDefaultScope
+
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		resource := p.ProtectedResource.String()
+		if !strings.HasSuffix(resource, "/") {
+			resource += "/"
+		}
+
+		if p.Scope != "" && !strings.HasPrefix(p.Scope, resource) {
+			p.Scope = resource + p.Scope
+		}
+		defaultScope = resource + defaultScope
+	}
+
+	p.ProviderData.setProviderDefaults(providerDefaults{
+		name:  adfsProviderName,
+		scope: defaultScope,
+	})
+
+	return &ADFSProvider{OIDCProvider: p}
```

Comment:
Related to method signature change suggestion above:

```suggestion
	return &ADFSProvider{
		OIDCProvider: &OIDCProvider{
			ProviderData: p,
		},
	}
```

Reply:
This is also assuming ADFS support the nonce part of the OIDC spec. If you need to disable that, you can also set `SkipNonce: true`